src/order/filter/bases.lean

```diff
@@ -0,0 +1,210 @@
+
+section same_type
+
+variables {l l' : filter α} {p : ι → Prop} {s : ι → set α} {t : set α} {i : ι}
+  {p' : ι' → Prop} {s' : ι' → set α} {i' : ι'}
+
+/-- Definition of `has_basis` unfolded to make it useful for `rw` and `simp`. -/
+lemma has_basis.mem_iff (hl : l.has_basis p s) : t ∈ l ↔ ∃ i (hi : p i), s i ⊆ t :=
+hl t
+
+lemma has_basis.mem_of_superset (hl : l.has_basis p s) (hi : p i) (ht : s i ⊆ t) : t ∈ l :=
+(hl t).2 ⟨i, hi, ht⟩
+
+lemma has_basis.mem_of_mem (hl : l.has_basis p s) (hi : p i) : s i ∈ l :=
+hl.mem_of_superset hi $ subset.refl _
+
+lemma has_basis.forall_nonempty_iff_ne_bot (hl : l.has_basis p s) :
+  (∀ {i}, p i → (s i).nonempty) ↔ l ≠ ⊥ :=
+⟨λ H, forall_sets_nonempty_iff_ne_bot.1 $
+  λ s hs, let ⟨i, hi, his⟩ := (hl s).1 hs in (H hi).of_subset his,
+  λ H i hi, inhabited_of_mem_sets H (hl.mem_of_mem hi)⟩
+
+lemma basis_sets (l : filter α) : l.has_basis (λ s : set α, s ∈ l) id :=
+λ t, exists_sets_subset_iff.symm
+
+lemma at_top_basis [nonempty α] [semilattice_sup α] :
+  (@at_top α _).has_basis (λ _, true) Ici :=
+λ t, by simpa only [exists_prop, true_and] using @mem_at_top_sets α _ _ t
+
+lemma at_top_basis' [semilattice_sup α] (a : α) :
+  (@at_top α _).has_basis (λ x, a ≤ x) Ici :=
+λ t, (@at_top_basis α ⟨a⟩ _ t).trans
+  ⟨λ ⟨x, _, hx⟩, ⟨x ⊔ a, le_sup_right, λ y hy, hx (le_trans le_sup_left hy)⟩,
+    λ ⟨x, _, hx⟩, ⟨x, trivial, hx⟩⟩
+
+theorem has_basis.ge_iff (hl' : l'.has_basis p' s')  : l ≤ l' ↔ ∀ i', p' i' → s' i' ∈ l :=
+⟨λ h i' hi', h $ hl'.mem_of_mem hi',
+  λ h s hs, let ⟨i', hi', hs⟩ := (hl' s).1 hs in mem_sets_of_superset (h _ hi') hs⟩
+
+theorem has_basis.le_iff (hl : l.has_basis p s) : l ≤ l' ↔ ∀ t ∈ l', ∃ i (hi : p i), s i ⊆ t :=
+by simp only [le_def, hl.mem_iff]
+
+theorem has_basis.le_basis_iff (hl : l.has_basis p s) (hl' : l'.has_basis p' s') :
+  l ≤ l' ↔ ∀ i', p' i' → ∃ i (hi : p i), s i ⊆ s' i' :=
+by simp only [hl'.ge_iff, hl.mem_iff]
+
+lemma has_basis.inf (hl : l.has_basis p s) (hl' : l'.has_basis p' s') :
+  (l ⊓ l').has_basis (λ i : ι × ι', p i.1 ∧ p' i.2) (λ i, s i.1 ∩ s' i.2) :=
+begin
+  intro t,
+  simp only [mem_inf_sets, exists_prop, hl.mem_iff, hl'.mem_iff],
+  split,
+  { rintros ⟨t, ⟨i, hi, ht⟩, t', ⟨i', hi', ht'⟩, H⟩,
+    use [(i, i'), ⟨hi, hi'⟩, subset.trans (inter_subset_inter ht ht') H] },
+  { rintros ⟨⟨i, i'⟩, ⟨hi, hi'⟩, H⟩,
+    use [s i, i, hi, subset.refl _, s' i', i', hi', subset.refl _, H] }
+end
+
+lemma has_basis.inf_principal (hl : l.has_basis p s) (s' : set α) :
+  (l ⊓ principal s').has_basis p (λ i, s i ∩ s') :=
+λ t, by simp only [mem_inf_principal, hl.mem_iff, subset_def, mem_set_of_eq,
+  mem_inter_iff, and_imp]
+
+lemma has_basis.eq_binfi (h : l.has_basis p s) :
+  l = ⨅ i (_ : p i), principal (s i) :=
+eq_binfi_of_mem_sets_iff_exists_mem $ λ t, by simp only [h.mem_iff, mem_principal_sets]
+
+lemma has_basis.eq_infi (h : l.has_basis (λ _, true) s) :
+  l = ⨅ i, principal (s i) :=
+by simpa only [infi_true] using h.eq_binfi
+
+@[nolint] -- Intentional use of `≥`
```

Comment:
Maybe refer to the Note `nolint_ge` instead, here and below.

Reply:
```suggestion
@[nolint] -- see Note [nolint_ge]
```